agent/app/agent_capability_test.go

```diff
@@ -0,0 +1,405 @@
+			dockerclient.Version_1_18,
+		}),
+		client.EXPECT().KnownVersions().Return([]dockerclient.DockerVersion{
+			dockerclient.Version_1_17,
+			dockerclient.Version_1_18,
+			dockerclient.Version_1_19,
+		}),
+	)
+
+	expectedCapabilityNames := []string{
+		"com.amazonaws.ecs.capability.privileged-container",
+		"com.amazonaws.ecs.capability.docker-remote-api.1.17",
+		"com.amazonaws.ecs.capability.docker-remote-api.1.18",
+		"com.amazonaws.ecs.capability.logging-driver.json-file",
+	}
+
+	var expectedCapabilities []*ecs.Attribute
+	for _, name := range expectedCapabilityNames {
+		expectedCapabilities = append(expectedCapabilities,
+			&ecs.Attribute{Name: aws.String(name)})
+	}
+
+	ctx, cancel := context.WithCancel(context.TODO())
+	// Cancel the context to cancel async routines
+	defer cancel()
+	agent := &ecsAgent{
+		ctx:                ctx,
+		cfg:                conf,
+		dockerClient:       client,
+		cniClient:          cniClient,
+		credentialProvider: aws_credentials.NewCredentials(mockCredentialsProvider),
+	}
+	capabilities := agent.capabilities()
+
+	for i, expected := range expectedCapabilities {
+		assert.Equal(t, aws.StringValue(expected.Name), aws.StringValue(capabilities[i].Name))
+		assert.Equal(t, aws.StringValue(expected.Value), aws.StringValue(capabilities[i].Value))
+	}
+
+	for _, capability := range capabilities {
+		if aws.StringValue(capability.Name) == "ecs.capability.task-eni-block-instance-metadata" {
+			t.Errorf("%s capability found when Task ENI is disabled in the config", taskENIBlockInstanceMetadataAttributeSuffix)
+		}
+	}
+}
+
+func TestCapabilitiesTaskResourceLimit(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	conf := &config.Config{TaskCPUMemLimit: true}
+
+	client := engine.NewMockDockerClient(ctrl)
+	versionList := []dockerclient.DockerVersion{dockerclient.Version_1_22}
+	client.EXPECT().SupportedVersions().Return(versionList)
+	client.EXPECT().KnownVersions().Return(versionList)
+
+	ctx, cancel := context.WithCancel(context.TODO())
+	// Cancel the context to cancel async routines
+	defer cancel()
+	agent := &ecsAgent{
+		ctx:          ctx,
+		cfg:          conf,
+		dockerClient: client,
+	}
+
+	expectedCapability := attributePrefix + capabilityTaskCPUMemLimit
+
+	capabilities := agent.capabilities()
+	capMap := make(map[string]bool)
+	for _, capability := range capabilities {
+		capMap[aws.StringValue(capability.Name)] = true
+	}
+
+	_, ok := capMap[expectedCapability]
+	assert.True(t, ok, "Should contain: "+expectedCapability)
+	assert.True(t, agent.cfg.TaskCPUMemLimit, "TaskCPUMemLimit should remain true")
+}
+
+func TestCapabilitesTaskResourceLimitDisabledByMissingDockerVersion(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	conf := &config.Config{TaskCPUMemLimit: true}
+
+	client := engine.NewMockDockerClient(ctrl)
+	versionList := []dockerclient.DockerVersion{dockerclient.Version_1_19}
+	client.EXPECT().SupportedVersions().Return(versionList)
```

Comment:
Same as above. It'd be a stronger test if we could enforce the ordering here as well.

Reply:
To me, it seems like checking for order is testing an implementation detail of the docker client instead of focusing on the contract between docker client and agent/app. I don't feel that strongly about this, though.